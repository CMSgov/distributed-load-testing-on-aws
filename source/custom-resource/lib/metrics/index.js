// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

const axios = require("axios");
const moment = require("moment");

const send = async (config, type) => {
  try {
    const metrics = {
      Solution: config.SolutionId,
      Version: config.Version,
      UUID: config.UUID,
      TimeStamp: moment().utc().format("YYYY-MM-DD HH:mm:ss.S"),
      Data: {
        Type: type,
        Region: config.Region,
        ExistingVpc: config.existingVPC,
      },
    };
    const params = {
      method: "post",
      port: 443,
      url: process.env.METRIC_URL,
      headers: {
        "Content-Type": "application/json",
      },
      data: metrics,
    };
    //Send Metrics & return status code.
    await axios(params);
  } catch (err) {
    //Not returning an error to avoid Metrics affecting the Application
    console.error(err);
  }
};

module.exports = {
  send: send,
};
