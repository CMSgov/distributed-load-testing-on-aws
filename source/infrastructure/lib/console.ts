// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { CloudFrontToS3 } from '@aws-solutions-constructs/aws-cloudfront-s3';
import { Effect, PolicyStatement } from '@aws-cdk/aws-iam';
import { Bucket, IBucket } from "@aws-cdk/aws-s3";
import { Construct, Tags } from '@aws-cdk/core';
import { Function as LambdaFunction } from '@aws-cdk/aws-lambda';


/**
 * @interface DLTConsoleContructProps
 * DLTConsoleContruct props
 */
export interface DLTConsoleContructProps {
    // Custom Resource Lambda
    customResource: LambdaFunction;
    // S3 Logs Bucket
    readonly s3LogsBucket: Bucket;
    // Solution ID
    readonly solutionId: string;
}

/**
 * Distributed Load Testing on AWS console construct
 * This creates the S3 bucket and CloudFront distribution 
 * and Cognito resources for the web front end.
 * @class
 */
export class DLTConsoleContruct extends Construct {
    public cloudFrontDomainName: string;
    public consoleBucketArn: string;
    public consoleBucket: IBucket;

    constructor(scope: Construct, id: string, props: DLTConsoleContructProps) {
        super(scope, id);

        const dltS3CloudFrontDist = new CloudFrontToS3(this, 'DLTCloudFrontToS3', {
            bucketProps: {
                serverAccessLogsBucket: props.s3LogsBucket,
                serverAccessLogsPrefix: 'console-bucket-access/',
            },
            cloudFrontDistributionProps: {
                comment: 'Website distribution for the Distributed Load Testing solution',
                enableLogging: true,
                errorResponses: [
                    { httpStatus: 403, responseHttpStatus: 200, responsePagePath: '/index.html' },
                    { httpStatus: 404, responseHttpStatus: 200, responsePagePath: '/index.html' }
                ],
                httpVersion: 'http2',
                logBucket: props.s3LogsBucket,
                logFilePrefix: 'cloudfront-logs/'
            },
            insertHttpSecurityHeaders: false
        });
        Tags.of(dltS3CloudFrontDist).add('SolutionId', props.solutionId);

        this.cloudFrontDomainName = dltS3CloudFrontDist.cloudFrontWebDistribution.domainName;
        this.consoleBucket = dltS3CloudFrontDist.s3BucketInterface;

        this.consoleBucketArn = dltS3CloudFrontDist.s3BucketInterface.bucketArn;

        props.customResource.addToRolePolicy(new PolicyStatement({
            effect: Effect.ALLOW,
            actions: ['s3:PutObject'],
            resources: [this.consoleBucketArn, `${this.consoleBucketArn}/*`]
        }));
    }
}