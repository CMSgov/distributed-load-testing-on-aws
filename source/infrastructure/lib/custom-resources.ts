// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { Aws, Construct, CustomResource } from '@aws-cdk/core';


/**
 * CustomResourcesConstruct props
 * @interface CustomResourcesConstructProps
 */
export interface CustomResourcesConstructProps {
    readonly apiEndpoint: string;
    //Custom Resource lambda
    readonly customResourceLambda: string;
    // Cognito Resources
    readonly cognitoIdentityPool: string;
    readonly cognitoUserPool: string;
    readonly cognitoUserPoolClient: string;
    // UI Console S3 bucket
    readonly consoleBucketName: string;
    // Test scenarios storage
    readonly scenariosBucket: string;
    // Source code details
    readonly sourceCodeBucketName: string;
    readonly sourceCodePrefix: string;
}

/**
 * @class
 * Distributed Load Testing on AWS Custom Resources Construct.
 * It creates a custom resource Lambda function, a solution UUID, and a custom resource to send anonymous usage.
 */
export class CustomResourcesConstruct extends Construct {

    constructor(scope: Construct, id: string, props: CustomResourcesConstructProps) {
        super(scope, id);

        new CustomResource(this, 'CopyConsoleFiles', {
            serviceToken: props.customResourceLambda,
            resourceType: 'Custom::CopyConsoleFiles',
            properties: {
                Resource: 'CopyAssets',
                SrcBucket: props.sourceCodeBucketName,
                SrcPath: `${props.sourceCodePrefix}/console`,
                ManifestFile: 'console-manifest.json',
                DestBucket: props.consoleBucketName
            }
        });

        const awsExports = `const awsConfig = {
            cw_dashboard: 'https://console.aws.amazon.com/cloudwatch/home?region=${Aws.REGION}#dashboards:name=',
            ecs_dashboard: 'https://${Aws.REGION}.console.aws.amazon.com/ecs/home?region=${Aws.REGION}#/clusters/${Aws.STACK_NAME}/tasks',
            aws_project_region: '${Aws.REGION}',
            aws_cognito_region: '${Aws.REGION}',
            aws_cognito_identity_pool_id: '${props.cognitoIdentityPool}',
            aws_user_pools_id: '${props.cognitoUserPool}',
            aws_user_pools_web_client_id: '${props.cognitoUserPoolClient}',
            oauth: {},
            aws_cloud_logic_custom: [
                {
                    name: 'dlts',
                    endpoint: '${props.apiEndpoint}',
                    region: '${Aws.REGION}'
                }
            ],
            aws_user_files_s3_bucket: '${props.scenariosBucket}',
            aws_user_files_s3_bucket_region: '${Aws.REGION}'
        }`;

        new CustomResource(this, 'ConsoleConfig', {
            serviceToken: props.customResourceLambda,
            resourceType: 'Custom::CopyConfigFiles',
            properties: {
                Resource: 'ConfigFile',
                DestBucket: props.consoleBucketName,
                AwsExports: awsExports
            }
        });
    }
}