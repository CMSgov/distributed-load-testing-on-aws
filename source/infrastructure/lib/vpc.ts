// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import {
    CfnSubnet,
    CfnVPC,
    CfnInternetGateway,
    CfnRouteTable,
    CfnVPCGatewayAttachment,
    CfnRoute,
    CfnSubnetRouteTableAssociation
} from '@aws-cdk/aws-ec2';
import { Aws, Construct, Fn, Tags } from '@aws-cdk/core';


/**
 * FargageVPCConstruct props
 * @interface FargateVpcContructProps
 */
export interface FargateVpcContructProps {
    // IP CIDR block for Subnet A
    readonly subnetACidrBlock: string;
    // IP CIDR block for Subnet B 
    readonly subnetBCidrBlock: string;
    // Solution ID
    readonly solutionId: string;
    // IP CIDR block for VPC 
    readonly vpcCidrBlock: string;
}

/**
 * @class
 * Distributed Load Testing on AWS VPC construct.
 * Creates a VPC for the Fagate test tasks. 
 * Includes 2 subnets across 2 availability zones and supporting resources.
 */
export class FargateVpcContruct extends Construct {
    // VPC
    public DLTfargateVpcId: string;
    public subnetA: string;
    public subnetB: string;

    constructor(scope: Construct, id: string, props: FargateVpcContructProps) {
        super(scope, id);

        const fargateVpc = new CfnVPC(this, 'DLTFargateVpc', {
            cidrBlock: props.vpcCidrBlock,
            enableDnsHostnames: true,
            enableDnsSupport: true
        });

        fargateVpc.addMetadata('cfn_nag', {
            rules_to_suppress: [{
                id: 'W60',
                reason: 'This VPC is used for the test runner Fargate tasks only, it does not require VPC flow logs.'
            }]
        });

        Tags.of(fargateVpc).add('SolutionId', props.solutionId);
        Tags.of(fargateVpc).add('Name', Aws.STACK_NAME)
        this.DLTfargateVpcId = fargateVpc.ref

        const subnetAResource = new CfnSubnet(this, 'DLTSubnetA', {
            cidrBlock: props.subnetACidrBlock,
            vpcId: this.DLTfargateVpcId,
            availabilityZone: Fn.select(0, Fn.getAzs())
        });
        Tags.of(subnetAResource).add('SolutionId', props.solutionId);
        this.subnetA = subnetAResource.ref

        const subnetBResource = new CfnSubnet(this, 'DLTSubnetB', {
            cidrBlock: props.subnetBCidrBlock,
            vpcId: this.DLTfargateVpcId,
            availabilityZone: Fn.select(1, Fn.getAzs())
        });
        Tags.of(subnetBResource).add('SolutionId', props.solutionId);
        this.subnetB = subnetBResource.ref

        const ig = new CfnInternetGateway(this, 'DLTFargateIG', {
        });
        Tags.of(ig).add('SolutionId', props.solutionId);

        const mainRouteTable = new CfnRouteTable(this, 'DLTFargateRT', {
            vpcId: this.DLTfargateVpcId,
        });
        Tags.of(mainRouteTable).add('SolutionId', props.solutionId);

        const gwa = new CfnVPCGatewayAttachment(this, 'DLTGatewayattachment', {
            vpcId: this.DLTfargateVpcId,
            internetGatewayId: ig.ref
        });

        const routeToInternet = new CfnRoute(this, 'DLTRoute', {
            destinationCidrBlock: '0.0.0.0/0',
            routeTableId: mainRouteTable.ref,
            gatewayId: ig.ref
        });
        routeToInternet.addDependsOn(gwa);

        new CfnSubnetRouteTableAssociation(this, 'DLTRouteTableAssociationA', {
            routeTableId: mainRouteTable.ref,
            subnetId: subnetAResource.ref
        });

        new CfnSubnetRouteTableAssociation(this, 'DLTRouteTableAssociationB', {
            routeTableId: mainRouteTable.ref,
            subnetId: subnetBResource.ref
        });

    }
}