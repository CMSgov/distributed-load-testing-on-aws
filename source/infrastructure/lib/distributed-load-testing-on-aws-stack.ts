// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import {
  Aspects,
  Aws,
  CfnCondition,
  CfnMapping,
  CfnParameter,
  CfnResource,
  Construct,
  Fn,
  IAspect,
  IConstruct,
  StackProps,
  Stack,
  CfnOutput
} from '@aws-cdk/core';
import { CognitoAuthConstruct } from './auth';
import { CommonResourcesContruct } from './common-resources';
import { FargateECSTestRunnerContruct } from './ecs';
import { FargateVpcContruct } from './vpc';
import { ScenarioTestRunnerStorageContruct } from './scenarios-storage';
import { DLTConsoleContruct } from './console';
import { CustomResourcesConstruct } from './custom-resources';
import { DLTAPI } from './api';
import { TestRunnerLambdasConstruct } from './test-task-lambdas';
import { TaskRunnerStepFunctionConstruct } from './step-functions';

/**
 * CDK Aspect implementation to set up conditions to the entire Construct resources
 */
class ConditionAspect implements IAspect {
  private readonly condition: CfnCondition;

  constructor(condition: CfnCondition) {
    this.condition = condition;
  }

  /**
   * Implement IAspect.visit to set the condition to whole resources in Construct.
   * @param {IConstruct} node Construct node to visit
   */
  visit(node: IConstruct): void {
    const resource = node as CfnResource;
    if (resource.cfnOptions) {
      resource.cfnOptions.condition = this.condition;
    }
  }
}

/**
 * Distributed Load Testing on AWS main CDK Stack
 * @class
 */
export class DLTStack extends Stack {
  // VPC ID
  private fargateVpcId: string;
  // Subnets for Fargate tasks
  private fargateSubnetA: string;
  private fargateSubnetB: string;

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    // CFN template format version
    this.templateOptions.templateFormatVersion = '2010-09-09';

    // CFN Parameters
    // Admin name
    const adminName = new CfnParameter(this, 'AdminName', {
      type: 'String',
      description: 'Admin user name to access the Distributed Load Testing console',
      minLength: 4,
      maxLength: 20,
      allowedPattern: '[a-zA-Z0-9-]+',
      constraintDescription: "Admin username must be a minimum of 4 characters and cannot include spaces"
    });

    // Admin email
    const adminEmail = new CfnParameter(this, 'AdminEmail', {
      type: 'String',
      description: 'Admin user email address to access the Distributed Load Testing Console',
      allowedPattern: '^[_A-Za-z0-9-\\+]+(\\.[_A-Za-z0-9-]+)*@[A-Za-z0-9-]+(\\.[A-Za-z0-9]+)*(\\.[A-Za-z]{2,})$',
      constraintDescription: 'Admin email must be a valid email address',
      minLength: 5
    });

    // Existing VPC ID
    const existingVpcId = new CfnParameter(this, 'ExistingVPCId', {
      type: 'String',
      description: 'Existing VPC ID',
      allowedPattern: '(?:^$|^vpc-[a-zA-Z0-9-]+)',
    });

    const existingSubnetA = new CfnParameter(this, 'ExistingSubnetA', {
      type: 'String',
      description: 'First existing subnet',
      allowedPattern: '(?:^$|^subnet-[a-zA-Z0-9-]+)'
    });

    const existingSubnetB = new CfnParameter(this, 'ExistingSubnetB', {
      type: 'String',
      description: 'Second existing subnet',
      allowedPattern: '(?:^$|^subnet-[a-zA-Z0-9-]+)'
    });

    // VPC CIDR Block
    const vpcCidrBlock = new CfnParameter(this, 'VpcCidrBlock', {
      type: 'String',
      default: '192.168.0.0/16',
      description: 'CIDR block of the new VPC where AWS Fargate will be placed',
      allowedPattern: '(?:^$|(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})/(\\d{1,2}))',
      constraintDescription: 'The VPC CIDR block must be a valid IP CIDR range of the form x.x.x.x/x.',
      minLength: 9,
      maxLength: 18
    });

    // Subnet A CIDR Block
    const subnetACidrBlock = new CfnParameter(this, 'SubnetACidrBlock', {
      type: 'String',
      default: '192.168.0.0/20',
      description: 'CIDR block for subnet A of the AWS Fargate VPC',
      allowedPattern: '(?:^$|(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})/(\\d{1,2}))',
      constraintDescription: 'The subnet CIDR block must be a valid IP CIDR range of the form x.x.x.x/x.',
      minLength: 9,
      maxLength: 18
    });

    // Subnet B CIDR Block
    const subnetBCidrBlock = new CfnParameter(this, 'SubnetBCidrBlock', {
      type: 'String',
      default: '192.168.16.0/20',
      description: 'CIDR block for subnet B of the AWS Fargate VPC',
      allowedPattern: '(?:^$|(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})/(\\d{1,2}))',
      constraintDescription: 'The subnet CIDR block must be a valid IP CIDR range of the form x.x.x.x/x.'
    });

    // Egress CIDR Block
    const egressCidrBlock = new CfnParameter(this, 'EgressCidr', {
      type: 'String',
      default: '0.0.0.0/0',
      description: 'CIDR Block to restrict the ECS container outbound access',
      minLength: 9,
      maxLength: 18,
      allowedPattern: '(?:^$|(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})/(\\d{1,2}))',
      constraintDescription: 'The Egress CIDR block must be a valid IP CIDR range of the form x.x.x.x/x.'
    });

    // CloudFormation metadata
    this.templateOptions.metadata = {
      'AWS::CloudFormation::Interface': {
        ParameterGroups: [
          {
            Label: { default: 'Console access' },
            Parameters: [adminName.logicalId, adminEmail.logicalId]
          },
          {
            Label: { default: 'Enter values here to use your own existing VPC' },
            Parameters: [existingVpcId.logicalId, existingSubnetA.logicalId, existingSubnetB.logicalId]
          },
          {
            Label: { default: 'Or have the solution create a new AWS Fargate VPC' },
            Parameters: [vpcCidrBlock.logicalId, subnetACidrBlock.logicalId, subnetBCidrBlock.logicalId, egressCidrBlock.logicalId]
          }
        ],
        ParameterLabels: {
          [adminName.logicalId]: { default: '* Console Administrator Name' },
          [adminEmail.logicalId]: { default: '* Console Administrator Email' },
          [existingVpcId.logicalId]: { default: 'The ID of an existing VPC in this region. Ex: `vpc-1a2b3c4d5e6f`' },
          [existingSubnetA.logicalId]: { default: 'The ID of a subnet within the existing VPC. Ex: `subnet-7h8i9j0k`' },
          [existingSubnetB.logicalId]: { default: 'The ID of a subnet within the existing VPC. Ex: `subnet-1x2y3z`' },
          [vpcCidrBlock.logicalId]: { default: 'AWS Fargate VPC CIDR Block' },
          [subnetACidrBlock.logicalId]: { default: 'AWS Fargate Subnet A CIDR Block' },
          [subnetBCidrBlock.logicalId]: { default: 'AWS Fargate Subnet A CIDR Block' },
          [egressCidrBlock.logicalId]: { default: 'AWS Fargate SecurityGroup CIDR Block' }
        }
      }
    };

    // CFN Mappings
    const solutionMapping = new CfnMapping(this, 'Solution', {
      mapping: {
        Config: {
          CodeVersion: 'CODE_VERSION',
          KeyPrefix: 'SOLUTION_NAME/CODE_VERSION',
          S3Bucket: 'CODE_BUCKET',
          SendAnonymousUsage: 'Yes',
          SolutionId: 'SO0062',
          URL: 'https://metrics.awssolutionsbuilder.com/generic'
        }
      }
    });
    const sendAnonymousUsage = solutionMapping.findInMap('Config', 'SendAnonymousUsage');
    const solutionId = solutionMapping.findInMap('Config', 'SolutionId');
    const solutionVersion = solutionMapping.findInMap('Config', 'CodeVersion');
    const sourceCodeBucket = Fn.join('-', [solutionMapping.findInMap('Config', 'S3Bucket'), Aws.REGION]);
    const sourceCodePrefix = solutionMapping.findInMap('Config', 'KeyPrefix');
    const metricsUrl = solutionMapping.findInMap('Config', 'URL');

    // CFN Conditions
    const sendAnonymousUsageCondition = new CfnCondition(this, 'SendAnonymousUsage', {
      expression: Fn.conditionEquals(sendAnonymousUsage, 'Yes')
    });

    const createFargateVpcResourcesCondition = new CfnCondition(this, 'CreateFargateVPCResources', {
      expression: Fn.conditionEquals(existingVpcId.valueAsString, '')
    });

    const usingExistingVpc = new CfnCondition(this, 'BoolExistingVPC', {
      expression: Fn.conditionNot(Fn.conditionEquals(existingVpcId.valueAsString, ''))
    });

    // Fargate VPC resources  
    const fargate = new FargateVpcContruct(this, 'DLTVpc', {
      solutionId: solutionId,
      subnetACidrBlock: subnetACidrBlock.valueAsString,
      subnetBCidrBlock: subnetBCidrBlock.valueAsString,
      vpcCidrBlock: vpcCidrBlock.valueAsString,
    });
    Aspects.of(fargate).add(new ConditionAspect(createFargateVpcResourcesCondition));
    this.fargateVpcId = Fn.conditionIf(createFargateVpcResourcesCondition.logicalId,
      fargate.DLTfargateVpcId,
      existingVpcId.valueAsString
    ).toString();

    this.fargateSubnetA = Fn.conditionIf(createFargateVpcResourcesCondition.logicalId,
      fargate.subnetA,
      existingSubnetA.valueAsString
    ).toString();

    this.fargateSubnetB = Fn.conditionIf(createFargateVpcResourcesCondition.logicalId,
      fargate.subnetB,
      existingSubnetB.valueAsString
    ).toString();

    // Fargate ECS resources
    const fargateECSResources = new FargateECSTestRunnerContruct(this, 'DLTEcs', {
      DLTfargateVpcId: this.fargateVpcId,
      securityGroupEgress: egressCidrBlock.valueAsString,
      solutionId: solutionId,
    });
    const existingVpc = Fn.conditionIf(usingExistingVpc.logicalId, true, false).toString();
    const commonResources = new CommonResourcesContruct(this, 'DLTCommonResources', {
      dltEcsTaskExecutionRole: fargateECSResources.dltTaskExecutionRole,
      solutionId: solutionId,
      sourceCodeBucket,
      sourceCodePrefix,
      solutionVersion,
      sendAnonymousUsageCondition,
      existingVpc,
      metricsUrl
    });

    const dltConsole = new DLTConsoleContruct(this, 'DLTConsoleResources', {
      customResource: commonResources.customResourceLambda,
      s3LogsBucket: commonResources.s3LogsBucket,
      solutionId: solutionId
    });

    const dltStorage = new ScenarioTestRunnerStorageContruct(this, 'DLTTestRunnerStorage', {
      ecsTaskExecutionRole: fargateECSResources.dltTaskExecutionRole,
      s3LogsBucket: commonResources.s3LogsBucket,
      cloudFrontDomainName: dltConsole.cloudFrontDomainName,
      solutionId,
    });

    const stepLambdaFunctions = new TestRunnerLambdasConstruct(this, 'DLTLambdaFunction', {
      cloudWatchLogsPolicy: commonResources.cloudWatchLogsPolicy,
      dynamoDbPolicy: dltStorage.dynamoDbPolicy,
      ecsTaskExecutionRoleArn: fargateECSResources.dltTaskExecutionRole.roleArn,
      ecsCloudWatchLogGroup: fargateECSResources.dltCloudWatchLogGroup,
      ecsCluster: fargateECSResources.dltEcsClusterName,
      ecsTaskDefinition: fargateECSResources.dltTaskDefinitionArn,
      ecsTaskSecurityGroup: fargateECSResources.dltSecurityGroupId,
      scenariosS3Policy: dltStorage.scenariosS3Policy,
      subnetA: this.fargateSubnetA,
      subnetB: this.fargateSubnetB,
      metricsUrl,
      sendAnonymousUsage,
      solutionId,
      solutionVersion,
      sourceCodeBucket: commonResources.sourceBucket,
      sourceCodePrefix,
      testScenariosBucket: dltStorage.scenariosBucket.bucketName,
      testScenariosTable: dltStorage.scenariosTable,
      uuid: commonResources.uuid,
    })

    const taskRunnerStepFunctions = new TaskRunnerStepFunctionConstruct(this, 'DLTStepFunction', {
      taskStatusChecker: stepLambdaFunctions.taskStatusChecker,
      taskRunner: stepLambdaFunctions.taskRunner,
      resultsParser: stepLambdaFunctions.resultsParser,
      taskCanceler: stepLambdaFunctions.taskCanceler,
      solutionId
    });

    const dltApi = new DLTAPI(this, 'DLTApi', {
      ecsCloudWatchLogGroup: fargateECSResources.dltCloudWatchLogGroup,
      cloudWatchLogsPolicy: commonResources.cloudWatchLogsPolicy,
      dynamoDbPolicy: dltStorage.dynamoDbPolicy,
      taskCancelerInvokePolicy: stepLambdaFunctions.taskCancelerInvokePolicy,
      scenariosBucketName: dltStorage.scenariosBucket.bucketName,
      scenariosS3Policy: dltStorage.scenariosS3Policy,
      scenariosTableName: dltStorage.scenariosTable.tableName,
      ecsCuster: fargateECSResources.dltEcsClusterName,
      ecsTaskExecutionRoleArn: fargateECSResources.dltTaskExecutionRole.roleArn,
      taskRunnerStepFunctionsArn: taskRunnerStepFunctions.taskRunnerStepFunctions.stateMachineArn,
      tastCancelerArn: stepLambdaFunctions.taskCanceler.functionArn,
      metricsUrl,
      sendAnonymousUsage,
      solutionId,
      solutionVersion,
      sourceCodeBucket: commonResources.sourceBucket,
      sourceCodePrefix,
      uuid: commonResources.uuid
    });

    const cognitoResources = new CognitoAuthConstruct(this, 'DLTCognitoAuth', {
      adminEmail: adminEmail.valueAsString,
      adminName: adminName.valueAsString,
      apiId: dltApi.apiId,
      cloudFrontDomainName: dltConsole.cloudFrontDomainName,
      scenariosBucketArn: dltStorage.scenariosBucket.bucketArn,
    });

    new CustomResourcesConstruct(this, 'DLTCustomResources', {
      apiEndpoint: dltApi.apiEndpointPath,
      customResourceLambda: commonResources.customResourceLambda.functionArn,
      cognitoIdentityPool: cognitoResources.cognitoIdentityPoolId,
      cognitoUserPool: cognitoResources.cognitoUserPoolId,
      cognitoUserPoolClient: cognitoResources.cognitoUserPoolClientId,
      consoleBucketName: dltConsole.consoleBucket.bucketName,
      scenariosBucket: dltStorage.scenariosBucket.bucketName,
      sourceCodeBucketName: commonResources.sourceBucket.bucketName,
      sourceCodePrefix
    });

    //Outputs
    new CfnOutput(this, 'Console', {
      description: 'Console URL',
      value: dltConsole.cloudFrontDomainName
    });
    new CfnOutput(this, 'SolutionUUID', {
      description: 'Solution UUID',
      value: commonResources.uuid
    })
  }
}
