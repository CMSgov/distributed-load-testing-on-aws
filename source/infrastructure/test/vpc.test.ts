// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from '@aws-cdk/core';
import { FargateVpcContruct } from '../lib/vpc';

test('DLT VPC Test', () => {
    const stack = new Stack();
    const vpc = new FargateVpcContruct(stack, 'TestVPC', {
        subnetACidrBlock: '10.0.0.0/24',
        subnetBCidrBlock: '10.0.1.0/24',
        solutionId: 'SO0062',
        vpcCidrBlock: '10.0.0.0/16',
    });

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
    expect(vpc.DLTfargateVpcId).toBeDefined();
    expect(stack).toHaveResource('AWS::EC2::VPC', {
        CidrBlock: '10.0.0.0/16',
        EnableDnsHostnames: true,
        EnableDnsSupport: true
    })
    expect(vpc.subnetA).toBeDefined();
    expect(stack).toHaveResource('AWS::EC2::Subnet', {
        CidrBlock: '10.0.0.0/24'
    });
    expect(vpc.subnetB).toBeDefined();
    expect(stack).toHaveResource('AWS::EC2::Subnet', {
        CidrBlock: '10.0.1.0/24'
    });
    expect(stack).toHaveResource('AWS::EC2::InternetGateway');
    expect(stack).toHaveResource('AWS::EC2::Route', {
        DestinationCidrBlock: '0.0.0.0/0',
        GatewayId: {
            Ref: "TestVPCDLTFargateIG4FFBAA11",
        },
        RouteTableId: {
            Ref: "TestVPCDLTFargateRT6952750D",
        }
    });
    expect(stack).toHaveResource('AWS::EC2::RouteTable');
    expect(stack).toHaveResource('AWS::EC2::VPCGatewayAttachment', {
        InternetGatewayId: {
            Ref: "TestVPCDLTFargateIG4FFBAA11",
        }
    });
    expect(stack).toHaveResource('AWS::EC2::SubnetRouteTableAssociation', {
        RouteTableId: {
            Ref: "TestVPCDLTFargateRT6952750D",
        },
        SubnetId: {
            Ref: "TestVPCDLTSubnetA8E320A43",
        }
    });
    expect(stack).toHaveResource('AWS::EC2::SubnetRouteTableAssociation', {
        RouteTableId: {
            Ref: "TestVPCDLTFargateRT6952750D",
        },
        SubnetId: {
            Ref: "TestVPCDLTSubnetB7A2BD254",
        }
    })
});