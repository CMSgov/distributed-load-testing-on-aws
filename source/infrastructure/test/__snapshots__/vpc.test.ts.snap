// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DLT VPC Test 1`] = `
Object {
  "Resources": Object {
    "TestVPCDLTFargateIG4FFBAA11": Object {
      "Type": "AWS::EC2::InternetGateway",
    },
    "TestVPCDLTFargateRT6952750D": Object {
      "Properties": Object {
        "VpcId": Object {
          "Ref": "TestVPCDLTFargateVpc0EC32C36",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "TestVPCDLTFargateVpc0EC32C36": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W60",
              "reason": "This VPC is used for the test runner Fargate tasks only, it does not require VPC flow logs.",
            },
          ],
        },
      },
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": Object {
              "Ref": "AWS::StackName",
            },
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "TestVPCDLTGatewayattachment7F5F62C2": Object {
      "Properties": Object {
        "InternetGatewayId": Object {
          "Ref": "TestVPCDLTFargateIG4FFBAA11",
        },
        "VpcId": Object {
          "Ref": "TestVPCDLTFargateVpc0EC32C36",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
    "TestVPCDLTRoute5972150C": Object {
      "DependsOn": Array [
        "TestVPCDLTGatewayattachment7F5F62C2",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "TestVPCDLTFargateIG4FFBAA11",
        },
        "RouteTableId": Object {
          "Ref": "TestVPCDLTFargateRT6952750D",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "TestVPCDLTRouteTableAssociationAD1563098": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "TestVPCDLTFargateRT6952750D",
        },
        "SubnetId": Object {
          "Ref": "TestVPCDLTSubnetA8E320A43",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "TestVPCDLTRouteTableAssociationB831818FD": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "TestVPCDLTFargateRT6952750D",
        },
        "SubnetId": Object {
          "Ref": "TestVPCDLTSubnetB7A2BD254",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "TestVPCDLTSubnetA8E320A43": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/24",
        "VpcId": Object {
          "Ref": "TestVPCDLTFargateVpc0EC32C36",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "TestVPCDLTSubnetB7A2BD254": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.1.0/24",
        "VpcId": Object {
          "Ref": "TestVPCDLTFargateVpc0EC32C36",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
  },
}
`;
